### PR TITLE
[LLM-eval] Adding context variables, input and multi output option

### PR DESCRIPTION
**Pull request by sunishsheth2009**
## What changes are proposed in this pull request?
[LLM-eval] Adding context variables, input and multi output option

## How is this patch tested?
- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/default_evaluator.py [hunk 1]
```diff
@@ -1011,14 +1050,37 @@ def _evaluate_custom_metrics(self, eval_df):
         for index, custom_metric in enumerate(self.custom_metrics):
             # deepcopying eval_df and builtin_metrics for each custom metric function call,
             # in case the user modifies them inside their function(s).
```
Comment (prithvikannan) [inline mlflow/models/evaluation/default_evaluator.py:1052]: can we please update this comment to explain this logic better

File: mlflow/models/evaluation/default_evaluator.py [hunk 2]
```diff
@@ -419,6 +420,44 @@ def _do_plot(**kwargs):
 }
 
 
+def _extract_output_and_other_columns(model_predictions, output_column_name):
```
Comment (prithvikannan) [inline mlflow/models/evaluation/default_evaluator.py:423]: in this function `model_predictions` comes from `self.model.predict`, which can return pandas.DataFrame, pandas.Series, numpy.ndarray or list ([docs](https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.PyFuncModel.predict)). can we make sure to handle all of these cases?

**Review by prithvikannan — CHANGES_REQUESTED** [review]

File: mlflow/models/evaluation/base.py [hunk 3]
```diff
@@ -85,10 +85,16 @@ def eval_fn(
...
+                    A Pandas or Spark DataFrame containing ``prediction``, ``target``,
+                    various ``variables``, ``input`` column.
                     The ``prediction`` column contains the predictions made by the model.
                     The ``target`` column contains the corresponding labels to the predictions made
                     on that row.
+                    The ``input`` column contains the input column for which the predictions
+                    were made.
+                    Each of the ``variables`` passed in make_metrics has a corresponding column
+                    with the same name as the variable name. These columns contain the
+                    corresponding variables values used for making the evaluations.
```
Comment (prithvikannan) [inline mlflow/models/evaluation/base.py:97]:
```suggestion
                    A Pandas or Spark DataFrame containing ``prediction``, ``target``,
                    various ``variables``, optional ``input`` column.
                    For each row...
                    - the ``prediction`` contains the prediction made by the model.
                    - the ``target`` contains the the expected output/label.
                    - the ``variables`` contain other outputs from the model that are used for making the evaluations.
                    - the optional ``input`` contains the input sent into the model.
```

File: mlflow/models/evaluation/base.py [hunk 4]
```diff
@@ -85,10 +85,16 @@ def eval_fn(
...
                 """
                 :param eval_df:
-                    A Pandas or Spark DataFrame containing ``prediction`` and ``target`` column.
+                    A Pandas or Spark DataFrame containing ``prediction``, ``target``,
+                    various ``variables``, ``input`` column.
                     The ``prediction`` column contains the predictions made by the model.
                     The ``target`` column contains the corresponding labels to the predictions made
                     on that row.
+                    The ``input`` column contains the input column for which the predictions
+                    were made.
```
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply mlflow/models/evaluation/base.py:94]: Spoke offline. We only consider the input for the LLM usecase which maps to the input column.

**Review by prithvikannan — APPROVED** [review]
LGTM, thanks @sunishsheth2009

**Pull request merged**